solc_select/solc_select.py

```diff
@@ -1,30 +1,139 @@
 import argparse
+import contextlib
 import hashlib
 import json
-from zipfile import ZipFile
 import os
-import shutil
+import platform
 import re
+import shutil
+import subprocess
 import sys
 import urllib.request
 from pathlib import Path
-from packaging.version import Version
+from urllib.parse import urlparse
+from zipfile import ZipFile
+
 from Crypto.Hash import keccak
+from packaging.version import Version
+
 from .constants import (
-    LINUX_AMD64,
-    MACOSX_AMD64,
-    WINDOWS_AMD64,
-    EARLIEST_RELEASE,
-    SOLC_SELECT_DIR,
     ARTIFACTS_DIR,
     CRYTIC_SOLC_ARTIFACTS,
     CRYTIC_SOLC_JSON,
+    EARLIEST_RELEASE,
+    LINUX_AMD64,
+    MACOSX_AMD64,
+    SOLC_SELECT_DIR,
+    WINDOWS_AMD64,
 )
 from .utils import mac_binary_is_universal, mac_can_run_intel_binaries
 
 Path.mkdir(ARTIFACTS_DIR, parents=True, exist_ok=True)
 
 
+def get_arch() -> str:
+    """Get the current system architecture."""
+    machine = platform.machine().lower()
+    if machine in ["x86_64", "amd64"]:
+        return "amd64"
+    elif machine in ["aarch64", "arm64"]:
+        return "arm64"
+    elif machine in ["i386", "i686"]:
+        return "386"
+    return machine
+
+
+def check_emulation_available() -> bool:
+    """Check if x86_64 emulation is available."""
+    if get_arch() != "arm64":
+        return False
+
+    # On macOS, check for Rosetta 2
+    if sys.platform == "darwin":
+        # Check if we can run x86_64 binaries
+        try:
+            result = subprocess.run(["arch", "-x86_64", "true"], capture_output=True, check=False)
+            return result.returncode == 0
+        except (FileNotFoundError, OSError):
+            return False
+
+    # On Linux, check for qemu-x86_64
+    try:
+        result = subprocess.run(
+            ["which", "qemu-x86_64"], capture_output=True, text=True, check=False
+        )
+        return result.returncode == 0
+    except (FileNotFoundError, OSError):
+        return False
+
+
+def get_emulation_prefix() -> list:
+    """Get the command prefix for emulation if needed."""
+    if get_arch() != "arm64":
+        return []
+
+    # On macOS, use arch command for Rosetta 2
+    if sys.platform == "darwin" and check_emulation_available():
+        return ["arch", "-x86_64"]
```

Comment:
this is not needed, you can just run the binary normally and rosetta will handle it automatically. Actually forcing it to run under x86_64 mode will cause a regression for newer solc binaries that are universal (intel+arm) by forcing them to run emulated when they could run natively.

Reply:
Excellent point! Removed the forced arch -x86_64 wrapper. Letting Rosetta handle universal binaries automatically is indeed the right approach and avoids forcing emulation when native execution is possible.